Code/src/main/java/nl/utwente/viskell/haskell/env/FunctionInfo.java

```diff
@@ -19,6 +33,19 @@ protected FunctionInfo(String name, Type signature) {
         this.signature = signature;
     }
 
+    public abstract Map<String, Object> toBundleFragment();
+
+    /** return a new instance of this type deserializing class-specific properties used in constructor **/
+    public static FunctionInfo fromBundleFragment(Map<String,Object> bundleFragment) throws NoSuchMethodException, InvocationTargetException, IllegalAccessException, ClassNotFoundException {
+        String kind = (String)bundleFragment.get(Bundleable.KIND);
+        String className = subClassMap.get(kind);
+        Class<?> clazz = Class.forName(className);
+
+        // Find the static "fromBundleFragment" method for the named type and call it
+        Method fromBundleMethod = clazz.getDeclaredMethod("fromBundleFragment", Map.class);
+        return (FunctionInfo)fromBundleMethod.invoke(null, bundleFragment);
+    }
+
```

Comment:
Hmm.

I believe (*believe*) the only FunctionInfo's that aren't from the catalog come from your own definitions, and that's just because those don't really have a category. CatalogFunction is the only subclass, and really the only special thing about catalog functions is that they have some extra strings (description, category). If we need to do this kind of magic to keep that distinction alive I think we might be better off collapsing the hierarchy, hoisting the extra strings up into FunctionInfo and making them optional (or even `Optional`).